pkg/storage/tsdb/tombstones.go

```diff
@@ -0,0 +1,360 @@
+	StartTime        int64                   `json:"start_time"`
+	EndTime          int64                   `json:"end_time"`
+	Selectors        []string                `json:"selectors"`
+	Matchers         []*labels.Matcher       `json:"-"`
+	UserID           string                  `json:"user_id"`
+	State            BlockDeleteRequestState `json:"state"`
+}
+
+func NewTombstone(userID string, requestTime int64, stateTime int64, startTime int64, endTime int64, selectors []string, requestID string, state BlockDeleteRequestState) *Tombstone {
+	return &Tombstone{
+		RequestCreatedAt: requestTime,
+		StateCreatedAt:   stateTime,
+		StartTime:        startTime,
+		EndTime:          endTime,
+		Selectors:        selectors,
+		UserID:           userID,
+		RequestID:        requestID,
+		State:            state,
+	}
+}
+
+// TombstoneManager is responsible for reading and writing tombstone files to the bucket.
+type TombstoneManager struct {
+	bkt    objstore.InstrumentedBucket
+	logger log.Logger
+}
+
+func NewTombstoneManager(
+	bkt objstore.Bucket,
+	userID string,
+	cfgProvider bucket.TenantConfigProvider,
+	logger log.Logger) *TombstoneManager {
+
+	return &TombstoneManager{
+		bkt:    bucket.NewUserBucketClient(userID, bkt, cfgProvider),
+		logger: util_log.WithUserID(userID, logger),
+	}
+}
+
+// Uploads a tombstone to object sotre
+func (m *TombstoneManager) WriteTombstone(ctx context.Context, tombstone *Tombstone) error {
+	data, err := json.Marshal(tombstone)
+	if err != nil {
+		return errors.Wrap(err, "serialize tombstone")
+	}
+
+	fullTombstonePath := path.Join(TombstonePath, getTombstoneFileName(tombstone.RequestID, tombstone.State))
+
+	// Check if the tombstone already exists for the same state. Could be the case the same request was made
+	// and is already in the middle of deleting series. Creating a new tombstone would restart
+	// the progress
+	tombstoneExists, err := m.TombstoneExists(ctx, tombstone.RequestID, tombstone.State)
+	if err != nil {
+		level.Error(m.logger).Log("msg", "unable to check if the same tombstone already exists", "requestID", tombstone.RequestID, "err", err)
+	} else if tombstoneExists {
+		return ErrTombstoneAlreadyExists
+	}
+
+	return errors.Wrap(m.bkt.Upload(ctx, fullTombstonePath, bytes.NewReader(data)), "upload tombstone")
+}
+
+func (m *TombstoneManager) TombstoneExists(ctx context.Context, requestID string, state BlockDeleteRequestState) (bool, error) {
+	fullTombstonePath := path.Join(TombstonePath, getTombstoneFileName(requestID, state))
+	exists, err := m.bkt.Exists(ctx, fullTombstonePath)
+
+	if exists || err != nil {
+		return exists, err
+	}
+
+	return false, nil
+}
+
+func (m *TombstoneManager) GetTombstoneByIDForUser(ctx context.Context, requestID string) (*Tombstone, error) {
+	found := []*Tombstone{}
+
+	for _, state := range AllDeletionStates {
+		filename := getTombstoneFileName(requestID, state)
+		exists, err := m.TombstoneExists(ctx, requestID, state)
```

Comment:
I wonder if we really need this extra check, which causes extra network call (I believe the `m.TombstoneExists` translates to a S3 HEAD http call).

Would it be possible to just do `m.ReadTombstone` and differentiate if the returned `err` is NotFound or other errors, then handle them differently?